src/tools/rbd/action/Export.cc

```diff
@@ -414,6 +414,36 @@ static int do_export_v2(librbd::Image& image, librbd::image_info_t &info, int fd
   ::encode(length, bl);
   ::encode(stripe_count, bl);
 
+  //encode imageMeta key and value
+  std::map<std::string, bufferlist> imagemetas;
+
+  r = image.metadata_list("", 0, &imagemetas);
```

Comment:
This needs to be a loop -- read X key/value pairs at a time and continue the loop if the number of retrieved pairs equals the request count.

Reply:
no need to be a loop in fact. `image.metadata_list("", 0, &imagemetas)` will fetch all key-value pairs into the map `imagemetas`  if we set the second parameter as 0.   Or am i missing sth ?

Reply:
It works for a large number of keys but not all possible (worse-case): https://github.com/ceph/ceph/blob/master/src/cls/rbd/cls_rbd.cc#L2654

Reply:
Hi @dillaman , if we specify the max_return to 0. I think the code you pointed is trying to read all metadatas in a loop until there is no ```more```.

Reply:
It returns a maximum of 64 keys if zero is provided

Reply:
if we set `max_return` as 0,  the `max_read` will be set to `RBD_MAX_KEYS_READ(64)`. but it only affect the the function `cls_cxx_map_get_vals` , the loop will break if no more image-meta ,which means that we get all metadatas.  so, when will the worse-case happen?

Reply:
If I tell cls_cxx_map_get_vals to retrieve a maximum of 5 entries and it returns 5 and says more are available, that's a bug. The more bool is supposed to be used to indicate a partial set being retrieved.

Reply:
@dillaman  I have a try to fix the bug you mentioned.
```
@@ -6418,13 +6419,15 @@ int PrimaryLogPG::do_osd_ops(OpContext *ctx, vector<OSDOp>& ops)
          tracepoint(osd, do_osd_op_pre_omapgetvals, soid.oid.name.c_str(), soid.snap.val, "???", 0, "???");
          goto fail;
        }
+
+       bool truncated = false;
        if (max_return > cct->_conf->osd_max_omap_entries_per_request) {
          max_return = cct->_conf->osd_max_omap_entries_per_request;
+         truncated = true;
        }
        tracepoint(osd, do_osd_op_pre_omapgetvals, soid.oid.name.c_str(), soid.snap.val, start_after.c_str(), max_return, filter_prefix.c_str());
 
        uint32_t num = 0;
-       bool truncated = false;
        bufferlist bl;
        if (oi.is_omap()) {
          ObjectMap::ObjectMapIterator iter = osd->store->get_omap_iterator(
@@ -6437,12 +6440,12 @@ int PrimaryLogPG::do_osd_ops(OpContext *ctx, vector<OSDOp>& ops)
          iter->upper_bound(start_after);
          if (filter_prefix > start_after) iter->lower_bound(filter_prefix);
          for (num = 0;
-              iter->valid() &&
+              num <= max_return &&
+                iter->valid() &&
                 iter->key().substr(0, filter_prefix.size()) == filter_prefix;
               ++num, iter->next(false)) {
            dout(20) << "Found key " << iter->key() << dendl;
-           if (num >= max_return ||
-               bl.length() >= cct->_conf->osd_max_omap_bytes_per_request) {
+           if (bl.length() >= cct->_conf->osd_max_omap_bytes_per_request) {
              truncated = true;
              break;
            }
```

But I have two questions about it:
(1) I am a little concern about changing this behavior of it if there is some other guys rely on this ```bug```. e.g ```rbd image-meta list```. maybe more. it would make a regression. 

(2) Actually, we can check the ```max_return``` and ```vals.size()``` to know did we only retrieved a partial set. And we can try again with a specified last_read. Then we don't need another parameter of ```more``` to tell us that. Just my opinion. 
Is there a design document for ```CEPH_OSD_OP_OMAPGETVALS``` op? 

Thanx

Reply:
(1) ```rbd image-meta list``` is broken and should be fixed [1]
(2) No such documentation -- the code is the documentation. The purpose is for case where the OSD cannot immediately return all entries but (1) more exist and (2) the ```max_return``` hasn't been met yet so the caller should try again (like receiving an -EAGAIN on a syscall).

[1] http://tracker.ceph.com/issues/21179

Reply:
Thanx @dillaman , will open a PR for this.

Reply:
@dillaman so, what should be done are as follow:  
1. fix the bug you mention, make `more` indicate a partial set being retrieved.
2. do some changes at where the `cls_cxx_map_get_vals` function is being used in `cls_rbd.cc` :  realize a new way to fetch all date instead of using `more` as a loop flag now

Reply:
I'm not sure I am following your last commit:
(1) the tracker ticket is to just update the rbd CLI to iterate over partial sets -- ```more``` already indicates that the OSD had to truncate the result but the requested limit has not been reached. See the ```osd_max_omap_bytes_per_request``` configuration value that can cause the truncation of the results.
(2) The ```more``` boolean should already be handled everywhere.

Reply:
hi @dillaman ,there are sth i am confused with. 
1. `image.metadata_list()`  is not broken currently. I have tried 300 metadates and succeed to get all.  (i think it is the 'bug' that makes it work)
2.  if we fix the bug you mentioned , which is that `more` should be `false` if we tell `cls_cxx_map_get_vals` to retrieve 5 entries and it returns 5 (am i right ?). But, please consider about this:  we have 65 metadates and want to get all.  The first time `cls_cxx_map_get_vals` succeeds to fetch 64 pairs(`RBD_MAX_KEYS_READ`) and `more` is set as `false`,  so it will not go to fetch the last pair, which means image.imagemeta_list() would be broken after fix. And that's why i said "realize a new logic to fetch all metadates instead of using more as a loop flag" in my last comment.
3. about "iterate over `image.metadata_list`": no matter how many time `image.metadata_list()` we call, the result of each time will be repeated.  `image.metadata_list()` will start from the ‘first‘ pair but not from the end of last function call to get metadates.

Reply:
OK -- looks like that was recently broken and is a totally separate issue. I'll push a fix for the broken handling of ```more``` [1].

[1] http://tracker.ceph.com/issues/21247